lib/handlers/authorize-handler.ts

```diff
@@ -0,0 +1,377 @@
+import * as url from 'url';
+import { AuthenticateHandler } from '.';
+import {
+  AccessDeniedError,
+  InvalidArgumentError,
+  InvalidClientError,
+  InvalidRequestError,
+  InvalidScopeError,
+  OAuthError,
+  ServerError,
+  UnauthorizedClientError,
+  UnsupportedResponseTypeError,
+} from '../errors';
+import { Client, Model, User } from '../interfaces';
+import { Request } from '../request';
+import { Response } from '../response';
+import { CodeResponseType, TokenResponseType } from '../response-types';
+import { hasOwnProperty } from '../utils/fn';
+import * as is from '../validator/is';
+
+/**
+ * Response types.
+ */
+
+const responseTypes = {
+  code: CodeResponseType,
+  token: TokenResponseType,
+};
+
+/**
+ * Constructor.
+ */
+
+export class AuthorizeHandler {
+  options: any;
+  allowEmptyState: boolean;
+  authenticateHandler: any;
+  model: Model;
+  constructor(options: any = {}) {
+    if (options.authenticateHandler && !options.authenticateHandler.handle) {
+      throw new InvalidArgumentError(
+        'Invalid argument: authenticateHandler does not implement `handle()`',
+      );
+    }
+
+    if (!options.model) {
+      throw new InvalidArgumentError('Missing parameter: `model`');
+    }
+
+    if (!options.model.getClient) {
+      throw new InvalidArgumentError(
+        'Invalid argument: model does not implement `getClient()`',
+      );
+    }
+
+    this.options = options;
+    this.allowEmptyState = options.allowEmptyState;
+    this.authenticateHandler =
+      options.authenticateHandler || new AuthenticateHandler(options);
+    this.model = options.model;
+  }
+
+  /**
+   * Authorize Handler.
+   */
+
+  async handle(request: Request, response: Response) {
+    if (!(request instanceof Request)) {
+      throw new InvalidArgumentError(
+        'Invalid argument: `request` must be an instance of Request',
+      );
+    }
+
+    if (!(response instanceof Response)) {
+      throw new InvalidArgumentError(
+        'Invalid argument: `response` must be an instance of Response',
+      );
+    }
+
+    if (request.query.allowed === 'false') {
+      throw new AccessDeniedError(
+        'Access denied: user denied access to application',
+      );
+    }
+
+    // Extend model object with request
+    this.model.request = request;
+
+    const client = await this.getClient(request);
+    const user = await this.getUser(request, response);
```

Comment:
In an implicit request that calls authorization. Does it make sense to call this.getUser here? Since the default implantation of getUser below uses the default AuthenticateHandler which expected access token in the request.

Reply:
@billxinli users should supply an `options.authenticateHandler` when using authorization grant type. see a sample in docs https://oauth2-server.readthedocs.io/en/latest/api/oauth2-server.html#authorize-request-response-options-callback

Reply:
Yeah tracing the code it seems one is able to pass in an optional `authenticateHandler`.

Wouldn't it be cleaner if it `getUser` was also implemented in the Model?

Seems weird to have to pass in a custom hander for implicit type.